# **Titanic Survival Prediction**

**Unveiling the Secrets of Survival: A Data-Driven Journey Through the Titanic Disaster**

This project dives deep into the historic Titanic dataset to predict which passengers survived the ill-fated voyage. Using machine learning, we analyze key factors like class, age, and gender to uncover the patterns of survival and build a predictive model. Join us as we explore the data, build a model, and gain insights into one of history's most compelling tragedies.

### **Step-1 : Importing required libraries**

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from warnings import filterwarnings
filterwarnings('ignore')

## **Step-2 : Loading the dataset**

Loading the pre-built **"Titanic"** dataset from the Seaborn library.

In [2]:
df = sns.load_dataset("titanic")

### **Step-3 : Data Overview**

In [3]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [4]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


### **Step-04 : Data Cleaning and Preprocessing**

Removing unwanted columns

In [6]:
df.drop(["deck", "embark_town", "adult_male", "alive", "class", "who"], axis=1, inplace=True)

In [7]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  alone
0         0       3    male  22.0      1      0   7.2500        S  False
1         1       1  female  38.0      1      0  71.2833        C  False
2         1       3  female  26.0      0      0   7.9250        S   True
3         1       1  female  35.0      1      0  53.1000        S  False
4         0       3    male  35.0      0      0   8.0500        S   True

Filling and Dropping missing values as applied

In [8]:
df['age'].fillna(df['age'].mean(), inplace=True)

In [9]:
df.dropna(subset=['embarked'], inplace=True)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  889 non-null    int64  
 1   pclass    889 non-null    int64  
 2   sex       889 non-null    object 
 3   age       889 non-null    float64
 4   sibsp     889 non-null    int64  
 5   parch     889 non-null    int64  
 6   fare      889 non-null    float64
 7   embarked  889 non-null    object 
 8   alone     889 non-null    bool   
dtypes: bool(1), float64(2), int64(4), object(2)
memory usage: 63.4+ KB


In [11]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  alone
0         0       3    male  22.0      1      0   7.2500        S  False
1         1       1  female  38.0      1      0  71.2833        C  False
2         1       3  female  26.0      0      0   7.9250        S   True
3         1       1  female  35.0      1      0  53.1000        S  False
4         0       3    male  35.0      0      0   8.0500        S   True

Label Encoding the columns where applicable

In [12]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['sex'] = le.fit_transform(df['sex'])
df['embarked'] = le.fit_transform(df['embarked'])

In [13]:
df = df.astype(int)

In [14]:
df.head()

survived  pclass  sex  age  sibsp  parch  fare  embarked  alone
0         0       3    1   22      1      0     7         2      0
1         1       1    0   38      1      0    71         0      0
2         1       3    0   26      0      0     7         2      1
3         1       1    0   35      1      0    53         2      0
4         0       3    1   35      0      0     8         2      1

### **Step-05 : Splitting the Training and Testing Data**

In [15]:
X = df.drop('survived', axis=1)
y = df['survived']

In [16]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### **Step-06 : Building a model**

Since its a **classification** problem, we will build a basic classification model, **Linear Regression**.

In [17]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [18]:
y_pred = model.predict(X_test)

### **Step-07 : Evaluating model**

Finally, we will evaluate our model using **confusion matrix, accuracy score, recall score, precision score and F1-score**.

Let's understand the metrics used to evaluate the model:

* **Confusion Matrix:** A table that summarizes the performance of a classification model. It shows the counts of true positives, true negatives, false positives, and false negatives.
* **Accuracy Score:** The proportion of correctly classified instances out of the total instances.
* **Recall Score (Sensitivity):** The proportion of actual positive instances that were correctly identified by the model.
* **Precision Score:** The proportion of predicted positive instances that were actually positive.
* **F1-Score:** The harmonic mean of precision and recall, providing a balanced measure of the model's performance.

In [20]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

accuracy = accuracy_score(y_test, y_pred)

accuracy

0.8033707865168539

In [21]:
confusion_matrix(y_test, y_pred)

array([[90, 19],
       [16, 53]])

In [23]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.83      0.84       109
           1       0.74      0.77      0.75        69

    accuracy                           0.80       178
   macro avg       0.79      0.80      0.79       178
weighted avg       0.81      0.80      0.80       178



### Model Performance Summary

Based on the evaluation metrics:

* **Accuracy:** The model achieved an accuracy of 80.34%, meaning it correctly predicted the survival status for about 80% of the test set.
* **Confusion Matrix:**
    * **True Positives (TP):** 53 (Correctly predicted survival)
    * **True Negatives (TN):** 90 (Correctly predicted non-survival)
    * **False Positives (FP):** 19 (Incorrectly predicted survival - Type I error)
    * **False Negatives (FN):** 16 (Incorrectly predicted non-survival - Type II error)
* **Classification Report:**
    * **Precision for survival (1):** 0.74 - When the model predicts survival, it is correct 74% of the time.
    * **Recall for survival (1):** 0.77 - The model correctly identifies 77% of the actual survivors.
    * **Precision for non-survival (0):** 0.85 - When the model predicts non-survival, it is correct 85% of the time.
    * **Recall for non-survival (0):** 0.83 - The model correctly identifies 83% of the actual non-survivors.
    * **F1-Score for survival (1):** 0.75 - A balanced score considering both precision and recall for survival.
    * **F1-Score for non-survival (0):** 0.84 - A balanced score considering both precision and recall for non-survival.

Overall, the model performs reasonably well, with slightly better performance in predicting non-survival compared to survival. The balance between precision and recall for both classes, as indicated by the F1-scores, suggests a decent trade-off in the model's predictions.